RandomArticleWidget/RandomArticleWidget.swift

```diff
@@ -0,0 +1,164 @@
+//
+//  RandomArticleWidget.swift
+//  RandomArticleWidget
+//
+//  Created by David Lynch on 2/5/19.
+//  Copyright © 2019 Wikimedia Foundation. All rights reserved.
+//
+
+import UIKit
+import NotificationCenter
+import WMF
+
+class RandomArticleWidget: UIViewController, NCWidgetProviding {
+    let randomArticleFetcher = RandomArticleFetcher()
+    let articlePreviewFetcher = WMFArticlePreviewFetcher()
+    let collapsedArticleView = ArticleRightAlignedImageCollectionViewCell()
+    let expandedArticleView = ArticleFullWidthImageCollectionViewCell()
+
+    var isExpanded = true
+    
+    var isEmptyViewHidden = true {
+        didSet {
+            collapsedArticleView.isHidden = !isEmptyViewHidden
+            expandedArticleView.isHidden = !isEmptyViewHidden
+        }
+    }
+    
+    var article: WMFArticle?
+//    var currentArticleKey: String?
+    
+    var dataStore: MWKDataStore? {
+        return SessionSingleton.sharedInstance()?.dataStore
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Do any additional setup after loading the view from its nib.
+        
+        view.translatesAutoresizingMaskIntoConstraints = false
+        
+        let tapGR = UITapGestureRecognizer(target: self, action: #selector(handleTapGesture(_:)))
+        view.addGestureRecognizer(tapGR)
+        
+        collapsedArticleView.frame = view.bounds
+        view.addSubview(collapsedArticleView)
+        
+        expandedArticleView.saveButton.addTarget(self, action: #selector(saveButtonPressed), for: .touchUpInside)
+        expandedArticleView.frame = view.bounds
+        view.addSubview(expandedArticleView)
+        
+        extensionContext?.widgetLargestAvailableDisplayMode = .expanded
+    }
+        
+    func widgetPerformUpdate(completionHandler: (@escaping (NCUpdateResult) -> Void)) {
+        // Perform any setup necessary in order to update the view.
+        
+        // If an error is encountered, use NCUpdateResult.Failed
+        // If there's no update required, use NCUpdateResult.NoData
+        // If there's an update, use NCUpdateResult.NewData
+        
+        defer {
+            updateView()
+        }
+        
+        let siteURL = SessionSingleton.sharedInstance().currentArticleSiteURL
+        self.randomArticleFetcher.fetchRandomArticle(withSiteURL: siteURL!, completion: { (error, result, summary) in
```

Comment:
For something like this it's better to `guard let siteURL = SessionSingleton.sharedInstance().currentArticleSiteURL else { ...failure case... return }` to prevent the force unwrap `!` which will crash if `siteURL` is nil